### PR TITLE
Kube client

### PR DESCRIPTION
**Pull request by aryszka**
partial implementation of #227 and https://github.com/zalando-incubator/kubernetes-on-aws/issues/169

(currently verbose debug logs left in the code to support testing of deployments)

### CODE REVIEW

File: dataclients/kube/kube.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+// logs if invalid, but proceeds with the valid ones
+// should report failures in Ingress status
+//
+// TODO:
+// - check how to set failures in ingress status
+func (c *Client) ingressToRoutes(items []*ingressItem) []*eskip.Route {
+	routes := make([]*eskip.Route, 0, len(items))
+	for _, i := range items {
+		if i.Metadata == nil || i.Metadata.Namespace == "" || i.Metadata.Name == "" {
+			log.Errorf("invalid ingress item: missing metadata")
```
Comment (mikkeloscar) [inline dataclients/kube/kube.go]: This item should be skipped if it's invalid right?

File: dataclients/kube/kube.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/zalando/skipper/eskip"
+)
+
+// FEATURE:
+// - provide option to limit the used namespaces?
+
+const (
+	defaultAPIAddress = "http://localhost:8001"
```
Comment (mikkeloscar) [inline dataclients/kube/kube.go]: Maybe it's a premature optimization, but I would make it possible to configure API access based on the environment when running in a cluster, so you don't have to have a 80+ MiB sidecar with `kubectl` just for this.

Example of how it's done in the kubernetes go client: https://github.com/kubernetes/client-go/blob/master/rest/config.go#L270-L298

I aslo made a PR for this in mate: https://github.com/zalando-incubator/mate/pull/21 :)

File: dataclients/kube/kube.go (hunk 1 above)
Comment (aryszka) [inline dataclients/kube/kube.go]: right

Comment (aryszka) [inline dataclients/kube/kube.go]: updated, thx

**Comment by mikkeloscar** [comment]
:+1:

File: skipper.go [hunk 3]
```diff
@@ -66,6 +53,10 @@ type Options struct {
 	// Skip TLS certificate check for etcd connections.
 	EtcdInsecure bool
 
+	// Kubernetes API URL for ingress. By default, Kubernetes ingress is
+	// disabled.
+	KubeUrl string
```
Comment (hjacobs) [inline skipper.go]: I would propose using the long form "kubernetes" everywhere (we don't need to save any ASCII chars, do we?)

File: dataclients/kube/definitions.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
+package kube
```
Comment (hjacobs) [inline dataclients/kube/definitions.go]: same here: I would propose using the long form "kubernetes" everywhere (we don't need to save any ASCII chars, do we?)

File: dataclients/kube/kube.go (hunk 1 above)
Comment (hjacobs) [inline dataclients/kube/kube.go]: not sure what your conventions say: but I would log this as warning and not error (as we will happily continue with the next item)

**Comment by hjacobs** [comment]
LGTM, see my comments regarding naming (I would prefer to use "kubernetes" in code, we are also not abbreviating "inkeeper" etc) and log level (stuff we can continue on is warning for me).

File: dataclients/kube/kube.go [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+	rsp, err := http.Get(url)
+	if err != nil {
+		log.Debugf("request to %s failed: %v", url, err)
+		return err
+	}
+
+	log.Debugf("request to %s succeeded", url)
+	defer rsp.Body.Close()
+
+	if rsp.StatusCode != 200 {
```
Comment (Raffo) [inline dataclients/kube/kube.go]: 200 -> http.StatusOK

File: skipper.go [hunk 6]
```diff
@@ -233,6 +224,10 @@ func createDataClients(o Options, auth innkeeper.Authentication) ([]routing.Data
 		clients = append(clients, etcdClient)
 	}
 
+	if o.KubeUrl != "" {
```
Comment (Raffo) [inline skipper.go]: Url -> URL

File: dataclients/kube/kube.go (hunk 1 above)
Comment (aryszka) [inline dataclients/kube/kube.go]: yes, true, changing it

File: dataclients/kube/definitions.go (hunk 4 above)
Comment (aryszka) [inline dataclients/kube/definitions.go]: for the other case, with the address option, i agree and changing right away. For the package name, i am a bit hesitant, but without a clear rational reasoning. Anyone with additional opinion on this.

File: dataclients/kube/kube.go (hunk 5 above)
Comment (aryszka) [inline dataclients/kube/kube.go]: right

File: skipper.go (hunk 6 above)
Comment (aryszka) [inline skipper.go]: yes.. should start to use this convention

File: skipper.go [hunk 7]
```diff
@@ -66,6 +53,10 @@ type Options struct {
 	// Skip TLS certificate check for etcd connections.
 	EtcdInsecure bool
 
+	// Kubernetes API base address for ingress. If not specififed,
+	// Kubernetes ingress is disabled.
+	KubernetesAPIAddress string
```
Comment (hjacobs) [inline skipper.go]: Why is it "Address" here and "Url" below? It should be Url here too I guess...

Comment (aryszka) [inline skipper.go]: i follow this naming convention to use address when it is only an address of a server, like http://localhost:8080 but no path or else. The innkeeper one just may be a full url, i don't remember. Is this dumb? Or are there better conventions for this distinction?

Comment (aryszka) [inline skipper.go]: maybe baseURL?

**Pull request merged**